java-checks/src/main/java/org/sonar/java/checks/SillyStringOperationsCheck.java

```diff
@@ -0,0 +1,183 @@
+          .addParameter(TypeCriteria.is(STRING))
+          .addParameter(TypeCriteria.is(STRING)),
+        MethodMatcher.create().typeDefinition(STRING).name("split")
+          .addParameter(TypeCriteria.is(STRING)),
+        MethodMatcher.create().typeDefinition(STRING).name("split")
+          .addParameter(TypeCriteria.is(STRING))
+          .addParameter("int"),
+        MethodMatcher.create().typeDefinition(STRING).name("startsWith")
+          .addParameter(TypeCriteria.is(STRING)),
+        MethodMatcher.create().typeDefinition(STRING).name("startsWith")
+          .addParameter(TypeCriteria.is(STRING))
+          .addParameter("int"),
+        MethodMatcher.create().typeDefinition(STRING).name("substring")
+          .addParameter("int"),
+        MethodMatcher.create().typeDefinition(STRING).name("substring")
+          .addParameter("int")
+          .addParameter("int"));
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree tree) {
+    if (tree.methodSelect().is(Kind.MEMBER_SELECT)) {
+      boolean issue;
+      ExpressionTree str = ((MemberSelectExpressionTree) tree.methodSelect()).expression();
+      Arguments args = tree.arguments();
+      String method = tree.symbol().name();
+      switch (method) {
+        case "contains":
+        case "compareTo":
+        case "compareToIgnoreCase":
+        case "endsWith":
+        case "indexOf":
+        case "lastIndexOf":
+        case "matches":
+        case "split":
+        case "startsWith":
+          issue = checkStartsWith(str, args);
+          break;
+        case "replaceAll":
+        case "replaceFirst":
+          issue = checkReplaceFirst(str, args);
+          break;
+        case "substring":
+          issue = checkSubstring(str, args);
+          break;
+        default:
+          issue = false;
+      }
+      if (issue) {
+        reportIssue(tree, format("Remove this \"%s\" call; it has predictable results.", method));
+      }
+    }
+  }
+
+  private static boolean checkStartsWith(ExpressionTree str, Arguments args) {
+    return isSameString(str, args.get(0));
+  }
+
+  private static boolean checkReplaceFirst(ExpressionTree str, Arguments args) {
+    return isSameString(str, args.get(0)) || isSameString(args.get(0), args.get(1));
+  }
+
+  private static boolean checkSubstring(ExpressionTree str, Arguments args) {
+    if (args.size() == 1) {
+      return isZero(args.get(0)) || isStringLength(str, args.get(0));
+    } else {
+      return isStringLength(str, args.get(0)) || (isZero(args.get(0)) && isStringLength(str, args.get(1)));
+    }
+  }
+
+  private static boolean isSameString(ExpressionTree str, ExpressionTree tree) {
+    return isSameSymbol(str, tree) || isSameStringLiteral(str, tree);
+  }
+
+  private static boolean isSameSymbol(ExpressionTree tree, ExpressionTree other) {
+    return Optional.ofNullable(symbol(tree)).map(s -> s.equals(symbol(other))).orElse(false);
```

Comment:
I think this usage of `Optional` is not adding much in terms of clarity. Using usual control flow constructs is more straightforward in Java